### ISSUE TITLE
Allow "generic" resource deployment to zeebe

### ISSUE THREAD

**Issue by marstamm** [camunda/camunda#20758]
**Is your feature request related to a problem? Please describe.**

We are working on a RPA integration for Camunda. For this, we need to manage the script files that the process has access to. Currently, only files with a specific type (`.bpmn`, `.dmn` and `.form`) can be deployed, anything else is rejected.

**Describe the solution you'd like**

Allow any file type to be part of the deployment and access it through the engine from a job worker:

* [ ] Any file can be part of a deployment, and deployed to the engine 
* [ ] A file local to the deployment can be linked and accessed from a job worker (https://github.com/camunda/camunda/issues/20758#issuecomment-2288799525)

**Describe alternatives you've considered**
 
- *Store script files outside of zeebe, e.g. on S3*: We want to present RPA as a core Camunda Feature. The RPA scripts will be the same for all instances of the process, so I think it should be handled the same as other linked resources, like DMN or Forms.
- *Only support `.robot` files instead of a generic resource*: This would work for our use-case, as only `.robot` files are supported. However, it makes it harder for users and customers to implement their custom solution, as they need to solve the problems for storage and referencing again. It would be great to have a simple solution for "versionBinding: deployment" for any resource.

**Additional context**

This was also discussed on Slack. This thread contains the discussion and some more background: https://camunda.slack.com/archives/CSQ2E3BT4/p1722330790414699

RPA Epic: https://github.com/camunda/product-hub/issues/2412

**Comment by megglos** [comment]
@aleksander-dytko as this is part of Boat, we should align on how to incorporate this. 

@marstamm what's deadline until you need this?

**Comment by marstamm** [comment]
It would be great if we could get this into 8.6, so we can present it at Camunda Con New York.

**Comment by megglos** [comment]
@marstamm we don't have capacity to include this into 8.6 still, is there any way you could unblock the demo for Camunda Con without this change in Zeebe?

cc @aleksander-dytko

**Comment by megglos** [comment]
@aleksander-dytko @bastiankoerber can you please align on a timeline for this going forward?

**Comment by bastiankoerber** [comment]
@marstamm & @nikku will follow up here

**Comment by marstamm** [comment]
We will be able to build RPA v0.1 without deployment to zeebe (cf. [Discussion](https://github.com/camunda/product-hub/issues/2412#issuecomment-2286041002)).

For v1.0, we might want this or a similar feature. We can postpone the discussion for this until then, potentially with a contribution from the RPA team

**Comment by nikku** [comment]
Some further insights into what [our investigation](https://github.com/camunda/product-hub/issues/2412#issuecomment-2286041002) found:

### User Problem

I want to deploy a resource with Zeebe, with the purpose to access what is deployed as part of my job worker.

### Assessment

This only works if we have a solid story how to access them from within a job worker in place, right now this story, to my knowledge, does not exist.

What folks would do today if we had "generic resource deployment" is to (1) either listen to Zeebe exporter stream for "resource deployed" events, or (2) ask operate for the resources (where in operate, this information may be available delayed).

### Potential Solution

We could offer a simple API (REST or zeebe) to retrieve a resource local to a deployment (`versionBinding=deployment`), or via key from the job worker.

### Example Scenario

* I deploy an external configuration (`.json` file) with my process
* I execute a job worker
* During execution I want to access the JSON file to understand "what to do"
* **I need to get from job -> process instance -> process definition -> deployment** and ask operate for a particular resource deployed with the process (?)
* I get access to the resource, and can follow-up
* :arrow_right: I'd like to be able to do this in a simple manner
* For certain _known_ and Zeebe managed resources with `versionBinding=latest` I may want to get access to the resource `key` that Zeebe assigned to it, so I can fetch exactly what I should.

**Comment by sbuettner** [comment]
@mathieu-stennier Would be great if you could list the AI/IDP use-cases here as well that would benefit from deployable configurations that could be picked up by other components like the `idp-connector`.

**Comment by sbuettner** [comment]
From the Connectors side we see quite some use-cases as well:
- External and re-usable Connector configuration: Instead of having all connection and technical connector details in the XML (bound via the Element Template) customers will be able to use configurations and reference them.
- Templates for text generation could be referenced by Connectors sending messages like EMaill, Sendgrid or AWS SNS
- Support custom decision engines via Connectors and support different standards (along with DMN) that can be picked up by Connectors

**Comment by mathieu-stennier** [comment]
@sbuettner 
Yes I would very much like to be able to deploy AI model configuration to a cluster and reference that model from AI powered connectors / Job workers.

![image](https://github.com/user-attachments/assets/9f5ab38d-3088-487b-8235-f0445262fc40)

Would that make sense?

**Comment by nikku** [comment]
> Yes I would very much like to be able to deploy AI model configuration to a cluster and reference that model from AI powered connectors / Job workers.

This would imply that we have "general configuration" available as a re-usable building block, served to external workers (such as RPA runners, in your case IDP runners), and subject to the existing versioning strategy (`latest`, `deployment`, `versionTag`).

**Comment by nikku** [comment]
> Under the hood, they could call an external job worker that runs scripts

Yes! :100:

**Comment by crobbins215** [comment]
@aleksander-dytko CC: @felix-mueller / @javi-alzorriz-camunda we have a dependency on this. @aleksander-dytko can you review this as we have a dependency from RPA on this https://github.com/camunda/product-hub/issues/2710#issuecomment-2912140827

**Comment by crobbins215** [comment]
> [@aleksander-dytko](https://github.com/aleksander-dytko) CC: [@felix-mueller](https://github.com/felix-mueller) / [@javi-alzorriz-camunda](https://github.com/javi-alzorriz-camunda) we have a dependency on this. [@aleksander-dytko](https://github.com/aleksander-dytko) can you review this as we have a dependency from RPA on this [camunda/product-hub#2710 (comment)](https://github.com/camunda/product-hub/issues/2710#issuecomment-2912140827)

@aleksander-dytko can you provide a response on this, it is a blocker for one of our epics and we need to understand how we can move forward.

**Comment by crobbins215** [comment]
Just spoke with @marstamm, we are moving ahead with a work around which means we will need to implement a migration when this has been completed https://github.com/camunda/product-hub/issues/2710#issuecomment-2912140827 please could you still confirm any target date for this.

**Comment by nikku** [comment]
Some thought experiment how this could be used/exposed through FEEL expressions to end-users, originating in [this conversation](https://camunda.slack.com/archives/C09QG3ZDRKJ/p1770909900730429?thread_ts=1770893454.084379&cid=C09QG3ZDRKJ):


#### Project setup

```
src/main/resources
|-> diagram.bpmn
|-> system-prompt.txt
```

#### `src/main/resources/diagram.bpmn`

```
input mapping:

* systemPrompt = camunda.resource[path=./system-prompt.txt]
```

#### `src/main/resources/system-prompt.txt`

```
You are a technocratic robot. Answer swiftly, with precision!
```

---

Taken above setup for re-use (by `id`) it gets complicated - 

#### `src/main/resources/diagram.bpmn`

```
input mapping:

* systemPrompt = camunda.resource[id="my-special-prompt"]
```

How to identify system prompt by ID, as it is a raw string? How to version it?

**Comment by eppdot** [comment]
After a discussion with @remcowesterhoud it became clear that fetching resources is currently implemented as a workaround (fetching a resources -> creates a fetch resource command within the engine -> responds with the resource stored in primary storage). 

With more generic files that are accessed much more frequently through the API this "workaround" implementation does not scale!

[Here](https://github.com/camunda/camunda/issues/43696) is another (tech-debt) issue that solves the query problem, but also there we need to make some more decisions
- should the generic/rpa files live at all in primary storage, if they are not accessed from within the engine? should they be stored through the engine at all?
- what metadata do we need to store in secondary storage? how do we clean up data? are there file size limits?
- how to deal with the latency between primary and secondary storage? does it matter?

**Comment by remcowesterhoud** [comment]
> what metadata do we need to store in secondary storage? how do we clean up data? are the file size limits?

We already store full `.bpmn`, `.dmn` and `.form` files. I don't see a problem adding generic resources to this. We already enforce a file size limit upon deployment, which will then be the same for these resources.

I think we can take inspiration from these existing resources on what metadata we need as well. Probably something like `key`, `resource`, `resourceName`, `version`, `versionTag`, `tenantId`. Maybe we should allow users to specify a type of resource as well?

Regarding cleanup, we don't support it for forms or RPA atm. We should add this. As resources could be long living we shouldn't have an automatic retention trigger on them.

> how to deal with the latency between primary and secondary storage? does it matter?

This latency is only relevant once upon deployment. I think it doesn't matter.

**Comment by sbuettner** [comment]
> Maybe we should allow users to specify a type of resource as well?
Yes, that would be important to be able to query resources by type especially when we use it for connection/configuration management cc @chillleader 

We might also want to allow querying the resource data from secondary storage.

**Comment by eppdot** [comment]
@sbuettner that would be the idea, to store (some - non bpmn/engine required) resources in secondary storage instead of primary storage (or more specifically its content + metadata). We could also think about deriving the "type" by default based on the file extension or allow searching resources (in secondary storage) by file extension, other fields, etc.

**Comment by sbuettner** [comment]
@eppdot We will probably need to have a more custom type mapping support. For example think about an external prompt file which lives in a *.md (markdown) file. It could be of type "system prompt" which wouldnt be easily extractable from the file extension suffix.

**Comment by sbuettner** [comment]
@nikku Do we already have a concept for adding additional information to process applications that could contain such information? For example in `.process-application` where one could declare such file type mappings/bindings?

**Comment by nikku** [comment]
> For example in .process-application where one could declare such file type mappings/bindings?

This, or a more general "camunda project" descriptor could be the place to establish such bindings - at least it should allow us to signal "these additional resources should be included", to prevent uploading an entire project into the engine ([`package.json#files`](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#files)).

We could combine this with pre-processors that parse meta-data, such as "versionTag", "id", etc. from the elements - a markdown file could expose such meta-data as part of a [YAML front-matter, for example](https://www.markdownlang.com/advanced/frontmatter.html#yaml-frontmatter):

```markdown
---
type: prompt
id: my-resource-id
versionTag: v100
---

You are a helpful AI agent...
```

IF we decide on core elements to include that should be versioned standalone (as I wrote earlier we don't need versioning for local includes), then the elements could also have `versionTag`, `id`, `camundaType` attached to themselves, as meta-data. Markdown shown above - 

```json
{
  "$meta": {
    "type": "...",
    "id": ...
  }
}
```

**Comment by sbuettner** [comment]
@nikku Yeah, I was just thinking that this should at least being discussed/considered in the context of https://github.com/camunda/product-hub/issues/3291

Supporting meta-data using custom pre-processors is nice but wont scale for generic file types we are not aware of (custom email template syntax, binary data,...). I think we should look into the project/meta-data file and then provide some custom OOTB pre-processors for typical file types.

**Comment by marcosgvieira** [comment]
🔄 **Status updated automatically (Core Features)**

**From:** Backlog  
**To:** Ready  

**Reason:** The issue already has an assignee.

**Comment by marcosgvieira** [comment]
🔄 **Team updated automatically (Core Features)**

**From:** core features charlie
**To:** core features bravo

**Reason:** The assignee of the issue is part of core features bravo

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** Apr - M2  
**To:** May - M1  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** May - M1  
**To:** May - M2  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** May - M2  
**To:** June - M1  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** June - M1  
**To:** June - M2  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by remcowesterhoud** [comment]
Closing this. The final documentation improvements are not fully related and are tracked by the docs team.
